Run express.js test suite as part of the tests

khardix opened the pull request:
This patch installs and runs express.js test suite on top of the base node image.

The goal is to check if the containerized Node expresses the same behavior as the upstream one. This should help ensure that even when i.e. patching Node to use older OpenSSL than officially supported, the result is still behaving as expected by an independent project.

Changes made in best-effort way; please kindly notify me if I reinvented wheel anywhere within the patch :slightly_smiling_face:.

hhorak commented:
[test]

hhorak commented on test/run:
```diff
@@ -6,13 +6,16 @@
 test -n $IMAGE_NAME \
   -a -n $VERSION
+# Change when new LTS version comes out
+readonly EXPRESS_REVISION="${EXPRESS_REVISION:-4.17.1}"
```
I'd suggest to use the latest released version, that can be retrieved by `npm show express version`.

khardix replied:
Replaced with `$(npm show express version)` as default, but keeping the `EXPRESS_REVISION` variable to allow overriding it.

hhorak commented:
Only a comment about hard-coding the `express` version that is IMO not necessary. Otherwise looks good to me.

hhorak commented:
CentOS 7 test fails because of `package nss_wrapper is not installed` error -- it used to be part of the mirror repos (http://vault.centos.org/7.6.1810/sclo/x86_64/rh/common/), but it is not any more. I think this needs to be fixed, because otherwise I don't know where else to get that package.

hhorak commented:
[tes]t

hhorak commented:
[test]

khardix commented:
[test]

khardix commented:
[test]

khardix commented:
It seems that npm is only installed in the container. In order to utilize `npm show express version`, I would then have to move the express preparation (git clone and config) into the running container; no obvious way how to do this jumps out at me. Any hints? :)

hhorak commented:
[test]

hhorak commented:
>  In order to utilize `npm show express version`, I would then have to move the express preparation

You can run only the `npm show...` in the container:
```
...
--branch "${EXPRESS_REVISION:-$(docker run --rm centos/nodejs-12-centos7 'npm show express version')}" --depth 1 \ 
...
```

khardix commented:
> You can run only the `npm show...` in the container

This one seems to work, thanks :slightly_smiling_face:  I have adjusted the express version query and squashed the fixup commits, so this should be ready for review again.

hhorak commented:
[test]

phracek commented on test/run:
```diff
@@ -57,25 +68,54 @@ run_s2i_build_proxy() {
+    --branch "${EXPRESS_REVISION}" --depth 1 \
+    'https://github.com/expressjs/express.git' "$1"
+  for key in "${!gitconfig[@]}"; do
+    git -C "$1" config --local "$key" "${gitconfig[$key]}"
```
Option `-C` does not exist in our testing `git` version. I would prefer to switch into `--work-dir` option.
```suggestion
    git --work-dir "$1" config --local "$key" "${gitconfig[$key]}"
```
https://git-scm.com/docs/git#Documentation/git.txt---work-treeltpathgt

phracek replied:
See error from tests:

https://gist.github.com/rhscl-bot/c1714c8f217a97bca53924c6c165a626#file-s2i-nodejs-container-179-sh-L3443

```bash
Running express.js test suite
Cloning into '/root/sources/test/express.js'...
Unknown option: -C
usage: git [--version] [--help] [-c name=value]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p|--paginate|--no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           <command> [<args>]
```

khardix replied:
Yeah, I noticed and used pushd/popd to change to the git root to avoid messing with extra CLI arguments… and then GitHub had a hiccup and the commit remained unpushed on my machine :slightly_smiling_face: Pushing it now.

hhorak commented:
[test] [keep-machine]

khardix commented:
[test] [keep-machine]

hhorak commented:
[test]
